torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1073,15 +1073,13 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
 
   if (work->recordFunctionEndCallback_) {
     // recordFunctionEndCallback_ is normally called in fininsh() function by
-    // base class, but since finish is not called by WorkNCCL, we schedule this
-    // function to be run when work is done.
+    // base class, but since finish is not called by WorkNCCL, we run this
+    // function now.
     // Note when can_profile is false, profilingTitle is not provided and so,
     // recordFunctionEndCallback_ is not set.
-    work->getFuture()->addCallback(std::move(work->recordFunctionEndCallback_));
+    work->recordFunctionEndCallback_();
```

Comment:
> The events that the FutureNCCL was using hadn't been recorded yet, thus blocking on them was a no-op. 

Does this mean the existing code for launching the cb was wrong?

Reply:
The code of FutureNCCL was correct (because, for NCCL, it's "correct" to always invoke callbacks inline).

The mistake was in ProcessGroupNCCL, which was constructing a FutureNCCL with "incomplete" arguments, but then using the FutureNCCL before making those arguments complete. (These "arguments" are the CUDA events).

Once FutureNCCL was returned by ProcessGroupNCCL, its arguments were complete, so the users of ProcessGroupNCCL couldn't hit this issue.

Reply:
I still believe that the high-level behavior here is wrong, since it caused recordFunctionEndCallback_ to be invoked _before_ the NCCL function was called. But this commit does nothing more than make that mistake explicit. I think it should be fixed in a separate PR by someone who knows what's going on.

Reply:
Yes, this is a bug and the reason why we disabled some of the failing tests as part of https://github.com/pytorch/pytorch/pull/48129. Attempting to fix/debug this as part of https://github.com/pytorch/pytorch/pull/48664, but in the meantime this PR should be fine as-is.